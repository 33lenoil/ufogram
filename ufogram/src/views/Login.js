import React, { useState } from 'react';
import {useNavigate } from 'react-router-dom';
import getUsers from '../api/user';
import axios from 'axios';
export default function Login() {
    // if account does not exist, retry or jump to registration page
    // if password incorrect, retry, block if wrong for 3 times

    const navigate = useNavigate();

    const [LoggedIn, setLoggedIn] = useState(false);
    const [error, setError] = useState(''); // State to store authentication error message

    let [username, setUsername] = useState('');
    let [password, setPassword] = useState('');
    let [userId, setUserId] = useState(0);
    
    const handleLogin = async () => {
        try {
            // Make an API call to your backend for authentication
            // const response = await axios.get('http://localhost:3000/Users', {
            //     method: 'GET',
            //     headers: {
            //         'Content-Type': 'application/json',
            //         'Cache-Control': 'no-cache',
            //         'Pragma': 'no-cache',
            //         'Expires': '0'
            //     },
            // });

            const response = await getUsers();

            if (response.status === 200) {

                for (let i = 0; i < response.data.length; i++) {
                   
                    if (response.data[i].username === username) {
                        setLoggedIn(true);
                        break;
                    }
                }
                if (LoggedIn) {
                    // setLoggedIn(true);
                    navigate('/main');
                } 
            } else {
                // Authentication failed, set error message
                setError('Invalid user list. Please try again.');
            }
        } catch (error) {
            console.error('Error during login:', error);
        }
    };

    const handleSignup = () => {
        navigate('/signup');
    };

    function handleUsernameChange(unameEvent) {
        setUsername(unameEvent.target.value);
    }

    function handlePasswordChange(passwdEvent) {
        setPassword(passwdEvent.target.value);
    };

    return (
        <div style={{display: "flex", justifyContent: "space-evenly"}}>
            <div style={{display: "flex", justifyContent: "center", position: "fixed", width: "100%", background: "#8769b6"}}>
                <h1>UFOgram</h1>
            </div>
            <div style={{display: "flex", width: "12%", marginTop: "100px", flexDirection: "column"}}>
                <label htmlFor='Username'>Username: </label>
                <input type="text" name="Username" value={username} onChange={handleUsernameChange}/>
                <label htmlFor='Password'>Password: </label>
                <input type="password" name="Password" value={password} onChange={handlePasswordChange}/>
                <button type="button" title="Log in" onClick={handleLogin}>Login</ button>
                <button type="button" title="Sign up" onClick={handleSignup} style={{color: "#808080"}}>Signup</ button>
            </div>
        </div>
    );
};


