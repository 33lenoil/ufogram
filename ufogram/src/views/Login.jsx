/* eslint-disable no-underscore-dangle */
import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import getUsers from '../api/user';

export default function Login() {
  const navigate = useNavigate();
  const [username, setUsername] = useState('zairuiy');
  const [password, setPassword] = useState('1234567');
  const [error, setError] = useState('');

  async function handleLogin() {
    try {
      const response = await getUsers();
      let userId = 0;

      if (response !== -1 && response.status === 200) {
        let userFound = false;
        for (let i = 0; i < response.data.users.length; i += 1) {
          if (response.data.users[i].username === username
            && response.data.users[i].password === password) {
            userFound = true;
            userId = response.data.users[i]._id;
            break;
          }
        }
        if (userFound) {
          navigate('/main', { state: { userId, username, users: response.data.users } });
        } else {
          setError('Invalid username or password');
        }
      } else {
        setError('Failed to fetch users');
      }
    } catch (loginError) {
      setError('Login error, please try again');
    }
  }

  const handleSignup = () => {
    navigate('/signup');
  };

  return (
    <div className="flex flex-col items-center justify-center min-h-screen bg-rose-quartz">
      <div className="bg-white p-8 rounded-lg shadow-lg w-80">
        <h1 className="text-3xl font-extrabold text-center mb-6 text-space-cadet" style={{ fontFamily: "'Comic Sans MS', cursive, sans-serif" }}>UFOgram</h1>
        {error && <p className="text-red-500 text-center">{error}</p>}
        <label htmlFor="usernameBox" className="block text-lg font-medium text-space-cadet mb-2">
          Username:
          <input
            type="text"
            name="Username"
            id="usernameBox"
            className="block w-full mt-1 p-2 border border-cadet-gray rounded-md shadow-sm focus:ring-space-cadet focus:border-space-cadet sm:text-sm"
            value={username}
            onChange={(e) => setUsername(e.target.value)}
          />
        </label>
        <label htmlFor="passwordBox" className="block text-lg font-medium text-space-cadet mb-4">
          Password:
          <input
            type="password"
            name="Password"
            data-testid="passwordBox"
            className="block w-full mt-1 p-2 border border-cadet-gray rounded-md shadow-sm focus:ring-space-cadet focus:border-space-cadet sm:text-sm"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
          />
        </label>

        <button
          type="button"
          title="Log in"
          className="w-full mt-4 p-2 bg-space-cadet text-white rounded-md shadow-sm transform transition-transform duration-300 hover:scale-105 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-space-cadet"
          onClick={handleLogin}
        >
          Login
        </button>
        <button
          type="button"
          title="Sign up"
          className="w-full mt-2 p-2 bg-mountbatten-pink text-white rounded-md shadow-sm transform transition-transform duration-300 hover:scale-105 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-mountbatten-pink"
          onClick={handleSignup}
        >
          Signup
        </button>
      </div>
    </div>
  );
}
