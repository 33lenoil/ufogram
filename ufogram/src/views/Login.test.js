/**
 * @jest-environment jsdom
 */

import React, { useState } from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router, Routes, Route } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import Login from './Login';

test('renders title', () => {
    render(
      <Router>
          {/* <Routes>
              <Route path="/signup" element={<Login />} />
          </Routes> */}
          <Login />
      </Router>
    );
    const linkElement = screen.getByText(/UFOgram/);
    expect(linkElement).toBeInTheDocument();
  });

test('renders username label', () => {
  render(
    <Router>
        {/* <Routes>
            <Route path="/signup" element={<Login />} />
        </Routes> */}
        <Login />
    </Router>
  );
  const linkElement = screen.getByText(/Username:/);
  expect(linkElement).toBeInTheDocument();
});

test('renders password label', () => {
    render(
        <Router>
            {/* <Routes>
                <Route path="/signup" element={<Login />} />
            </Routes> */}
            <Login />
        </Router>
      );
      const linkElement = screen.getByText(/Password:/);
    expect(linkElement).toBeInTheDocument();
});

test('renders signup button', () => {
    render(
        <Router>
            {/* <Routes>
                <Route path="/signup" element={<Login />} />
            </Routes> */}
            <Login />
        </Router>
      );
  const linkElement = screen.getByRole('button', {
    name: /Signup/
  })
  expect(linkElement).toBeInTheDocument();
});

test('renders login button', () => {
    render(
        <Router>
            {/* <Routes>
                <Route path="/signup" element={<Login />} />
            </Routes> */}
            <Login />
        </Router>
      );
  const linkElement = screen.getByRole('button', {
    name: /Login/
  })
  expect(linkElement).toBeInTheDocument();
});

/**
 * Snapshot Testing
 */

test('the component matches the snapshot', () => {
  const component = renderer.create(<Router>
    {/* <Routes>
        <Route path="/signup" element={<Signup />} />
    </Routes> */}
    <Login />
</Router>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

/**
 *  Testing
 */

// test('welcome is displayed after click the registration button', async () => {
//   render(<Login />);
//   const buttonElement = screen.getByRole('button');
//   await userEvent.click(buttonElement);
//   const welcomeElement = screen.getAllByLabelText(/welcome/i);
//   expect(welcomeElement).toBeInTheDocument();
// });

// test('Username is displayed after clicking on registration button', async () => {
//   render(<Login />);
//   const usernameElement = screen.getByTestId('uname');
//   await userEvent.type(usernameElement, 'lili');
//   const buttonElement = screen.getByRole('button');
//   await userEvent.click(buttonElement);
//   const welcomeElement = screen.getAllByText(/lili/i);
//   expect(welcomeElement).toBeInTheDocument();
// });
