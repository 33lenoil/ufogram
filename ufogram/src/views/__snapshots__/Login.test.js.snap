// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the component matches the snapshot 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "justifyContent": "space-evenly",
    }
  }
>
  <div
    style={
      Object {
        "background": "#8769b6",
        "display": "flex",
        "justifyContent": "center",
        "position": "fixed",
        "width": "100%",
      }
    }
  >
    <h1>
      UFOgram
    </h1>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "marginTop": "100px",
        "width": "12%",
      }
    }
  >
    <label
      htmlFor="Username"
    >
      Username: 
    </label>
    <input
      id="usernameBox"
      name="Username"
      onChange={[Function]}
      type="text"
    />
    <label
      htmlFor="Password"
    >
      Password: 
    </label>
    <input
      data-testid="passwordBox"
      name="Password"
      onChange={[Function]}
      type="password"
    />
    <button
      onClick={[Function]}
      title="Log in"
      type="button"
    >
      Login
    </button>
    <button
      onClick={[Function]}
      style={
        Object {
          "color": "#808080",
        }
      }
      title="Sign up"
      type="button"
    >
      Signup
    </button>
  </div>
</div>
`;
